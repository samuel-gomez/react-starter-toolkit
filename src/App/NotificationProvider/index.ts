export { default, NotificationContext } from './NotificationProvider';
export { default as useNotifications } from './Notifications.hook';
export type { TReturnUseNotification } from './Notifications.hook';
export type { TaddNotification, TonDeleteNotification } from './Notifications.hook';
export type { EType } from './Notifications';
