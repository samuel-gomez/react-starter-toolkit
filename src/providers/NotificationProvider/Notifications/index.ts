export { default } from './Notifications.container';
export type { TNotifications } from './Notifications';
export type { TNotification, EType } from './Notification';
