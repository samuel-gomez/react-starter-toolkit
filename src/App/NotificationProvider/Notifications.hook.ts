import { useState, useCallback } from 'react';
import { MODIFIER_HIDE, DELAY_HIDE } from './Notifications/constants';
import { TNotification } from './Notifications/Notification';

export type TaddNotification = (notification: TNotification) => void;
export type TonDeleteNotification = ((id: string) => void) | null;

export type TReturnUseNotification = {
  addNotification: TaddNotification;
  onDeleteNotification: TonDeleteNotification;
  clearAllNotifications: (() => void) | null;
  stateNotifications: TNotification[];
};

const useNotifications = (initState?: TNotification[]): TReturnUseNotification => {
  const [stateNotifications, setStateNotifications] = useState(initState || []);

  const clearAllNotifications = useCallback(() => {
    setStateNotifications([]);
  }, []);

  const onDeleteNotification = useCallback(
    (id: string) => {
      const notificationsWithHide = stateNotifications.map(notification =>
        notification.id === id ? { ...notification, classModifier: MODIFIER_HIDE } : notification,
      );

      setStateNotifications(notificationsWithHide);
      setTimeout(() => {
        const filteredNotifications = stateNotifications.filter(notification => notification.id !== id);
        setStateNotifications(filteredNotifications);
      }, DELAY_HIDE);
    },
    [stateNotifications],
  );

  const addNotification = useCallback(
    (notification: TNotification) => {
      if (stateNotifications.filter(notif => notif.id === notification.id).length === 0) {
        setStateNotifications([...stateNotifications, notification]);
      }
    },
    [stateNotifications],
  );

  return { addNotification, onDeleteNotification, clearAllNotifications, stateNotifications };
};

export default useNotifications;
