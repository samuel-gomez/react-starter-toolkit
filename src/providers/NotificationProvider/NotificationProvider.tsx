import { createContext, ReactNode, useMemo } from 'react';
import { emptyFunction } from 'shared/helpers';
import Notifications from './Notifications';
import useNotifications, { TaddNotification } from './Notifications.hook';

export type TNotificationContext = {
  addNotification: TaddNotification;
};

export const NotificationContext = createContext<TNotificationContext>({ addNotification: emptyFunction });
NotificationContext.displayName = 'NotificationContext';

type TNotificationProvider = {
  children: ReactNode;
  useNotificationsFn?: typeof useNotifications;
};

const NotificationProvider = ({ children, useNotificationsFn = useNotifications }: TNotificationProvider) => {
  const { addNotification, onDeleteNotification, stateNotifications } = useNotificationsFn();
  const value = useMemo(() => ({ addNotification }), [addNotification]);

  return (
    <NotificationContext.Provider value={value}>
      {stateNotifications && stateNotifications.length > 0 && (
        <Notifications notifications={stateNotifications} deleteNotification={onDeleteNotification} />
      )}
      {children}
    </NotificationContext.Provider>
  );
};

export default NotificationProvider;
