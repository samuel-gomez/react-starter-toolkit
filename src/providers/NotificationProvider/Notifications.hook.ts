import { useState, useCallback } from 'react';
import { MODIFIER_HIDE, DELAY_HIDE } from './Notifications/constants';
import { TNotification } from './Notifications/Notification';

const useNotifications = (initState: TNotification[] = []) => {
  const [stateNotifications, setStateNotifications] = useState(initState);

  const clearAllNotifications = useCallback(() => {
    setStateNotifications([]);
  }, []);

  const onDeleteNotification = useCallback(
    (id: string) => {
      const notificationsWithHide = stateNotifications.map(notification =>
        notification.id === id ? { ...notification, classModifier: MODIFIER_HIDE } : notification,
      );

      setStateNotifications(notificationsWithHide);
      setTimeout(() => {
        const filteredNotifications = stateNotifications.filter(notification => notification.id !== id);
        setStateNotifications(filteredNotifications);
      }, DELAY_HIDE);
    },
    [stateNotifications],
  );

  const addNotification = useCallback(
    (notification: TNotification) => {
      if (stateNotifications.filter(notif => notif.id === notification.id).length === 0) {
        setStateNotifications([...stateNotifications, notification]);
      }
    },
    [stateNotifications],
  );

  return { addNotification, onDeleteNotification, clearAllNotifications, stateNotifications };
};

export type TReturnUseNotification = ReturnType<typeof useNotifications>;
export type TaddNotification = TReturnUseNotification['addNotification'];
export type TonDeleteNotification = TReturnUseNotification['onDeleteNotification'];

export default useNotifications;
