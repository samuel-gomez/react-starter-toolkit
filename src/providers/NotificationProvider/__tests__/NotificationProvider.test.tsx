import { useContext } from 'react';
import { render } from '@testing-library/react';
import NotificationProvider, { NotificationContext } from '../NotificationProvider';
import type { TNotificationContext } from '../NotificationProvider';
import useNotifications from '../Notifications.hook';

const Base = ({ addNotification }: TNotificationContext) => {
  addNotification({ id: '02', label: 'title2' });
  return (
    <ul>
      <li>have addNotification</li>
    </ul>
  );
};

const BaseWithNotification = () => {
  const notificationProps = useContext(NotificationContext);
  return <Base {...notificationProps} />;
};
const addNotificationMock = jest.fn();
const onDeleteNotificationMock = jest.fn();

const App = ({ useNotificationsFn }: { useNotificationsFn: typeof useNotifications }) => (
  <NotificationProvider useNotificationsFn={useNotificationsFn}>
    <BaseWithNotification />
  </NotificationProvider>
);

describe('Render App with Base have addNotification props', () => {
  it('Should Base have fetchCustom props when call setFetchCustom', () => {
    const useNotificationsMock = jest.fn().mockReturnValue({
      addNotification: addNotificationMock,
      onDeleteNotification: onDeleteNotificationMock,
      stateNotifications: [],
    });
    const { asFragment, getByText } = render(<App useNotificationsFn={useNotificationsMock} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('have addNotification')).toBeDefined();
  });

  it('Should Base have fetchCustom props when call setFetchCustom with 1 notification', () => {
    const notification = {
      code: 200,
      detail: 'detail',
      label: 'labeldemonalert',
      id: 'id',
    };

    const useNotificationsMock = jest.fn().mockReturnValue({
      addNotification: addNotificationMock,
      onDeleteNotification: onDeleteNotificationMock,
      stateNotifications: [notification],
    });

    const { asFragment, getByText, getByRole } = render(<App useNotificationsFn={useNotificationsMock} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('have addNotification')).toBeDefined();
    expect(getByRole('alert')).toBeDefined();
    expect(getByText('labeldemonalert')).toBeDefined();
  });
});
