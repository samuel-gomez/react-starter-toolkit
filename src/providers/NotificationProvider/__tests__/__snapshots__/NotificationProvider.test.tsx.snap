// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App with Base have addNotification props Should Base have fetchCustom props when call setFetchCustom 1`] = `
<DocumentFragment>
  <ul>
    <li>
      have addNotification
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Render App with Base have addNotification props Should Base have fetchCustom props when call setFetchCustom with 1 notification 1`] = `
<DocumentFragment>
  <aside
    class="af-notifications af-notifications--open"
  >
    <div
      class="af-alert af-alert--notification af-alert--error"
      role="alert"
    >
      <div
        class="af-alert__title"
      >
        <div
          class="af-alert__title-icon"
        >
          <i
            class="glyphicon glyphicon-minus-sign"
          />
        </div>
        <div
          class="af-alert__title-text"
        >
          labeldemonalert
        </div>
        <button
          class="af-alert__title-icon-close"
          type="button"
        >
          <i
            class="glyphicon glyphicon-close"
          />
        </button>
      </div>
      <div
        class="af-alert__content"
      >
        <div
          class="af-alert__content__left"
        />
        <div
          class="af-alert__content__right"
        >
          <p>
            detail
          </p>
        </div>
      </div>
    </div>
  </aside>
  <ul>
    <li>
      have addNotification
    </li>
  </ul>
</DocumentFragment>
`;
